File: braccio_tp/braccio_tp/student_work/transformation.py
import numpy as np
from math import sin, cos, pi

class Transformation:
    def __init__(self, rotation=np.eye(3), position=np.zeros(3)):
        """
        Initialise une transformation avec une matrice de rotation (3x3) et un vecteur de position (3x1).
        Par défaut, la rotation est l'identité et la position est le vecteur nul.
        """
        self.rotation = rotation  # Matrice 3x3
        self.position = position  # Vecteur 3x1


    def set_translation(self, tx, ty, tz):
        self.position[0] = tx
        self.position[1] = ty
        self.position[2] = tz

    def set_rotation(self, roll, pitch, yaw):
        self.rotation = RotationX(roll) # * ...

    def set_rotation_from_quaternion(self,x,y,z,w):
        self.rotation[0,0] = 1.0
        self.rotation[0,1] = 0.0
        self.rotation[0,2] = 0.0
        self.rotation[1,0] = 0.0
        self.rotation[1,1] = 1.0
        self.rotation[1,2] = 0.0
        self.rotation[2,0] = 0.0
        self.rotation[2,1] = 0.0
        self.rotation[2,2] = 1.0


    def RotationX(theta):
        """Crée la matrice de rotation  pour une rotation autour de l'axe X."""
        rotation = np.array([
            [1, 0, 0],
            [0, cos(theta), -sin(theta)],
            [0, sin(theta), cos(theta)]
        ])
        return rotation

    def RotationY(theta):
        """Crée la matrice de rotation  pour une rotation autour de l'axe Y."""
        rotation = np.array([
            [cos(theta), 0, sin(theta)],
            [0, 1, 0],
            [-sin(theta), 0, cos(theta)]
        ])
        return rotation

    def RotationZ(theta):
        """Crée la matrice de rotation  pour une rotation autour de l'axe Z."""
        rotation = np.array([
            [cos(theta), -sin(theta), 0],
            [sin(theta), cos(theta), 0],
            [0, 0, 1]
        ])
        return rotation

    def multiplication(self, other):
        """Compose cette transformation avec une autre (self * other)."""
        """Attention le signe * sur les np.array fait une multiplication terme à terme """
        """la multiplication matricielle se fait avec l'opérateur @ """

        new_rotation = np.dot(self.rotation, other.rotation)
        new_position = self.position + np.dot(self.rotation, other.position)

        return Transformation(new_rotation, new_position)

    def inverse(self):
        """Retourne l'inverse de cette transformation."""
        inv_rotation = self.rotation.T
        inv_position = -np.dot(inv_rotation, self.position)

        return Transformation(inv_rotation, inv_position)

    def __mul__(self, other):
        """Définit l'opérateur * pour composer deux transformations (self * other)."""
        return self.multiplication(other)

    def __str__(self):
        """Affiche la transformation sous forme lisible."""
        return f"Rotation:\n{self.rotation}\nPosition: {self.position}"

File: braccio_tp/student/Transformation.cpp
#include "Transformation.h"

// retourne la matrice de rotation en fonction des quaternions
Eigen::Matrix<double,3,3> setQuaternion( const double & a, const double & b, const double & c, const double &d)
{
    Eigen::Matrix<double,3,3> rotation;

    rotation(0,0) = a*a + b*b - c*c - d*d;
    rotation(0,1) = 2*(b*c - a*d);
    rotation(0,2) = 2*(b*d + a*c);

    rotation(1,0) = 2*(b*c + a*d);
    rotation(1,1) = a*a - b*b + c*c - d*d;
    rotation(1,2) = 2*(c*d - a*b);

    rotation(2,0) = 2*(b*d - a*c);
    rotation(2,1) = 2*(c*d + a*b);
    rotation(2,2) = a*a - b*b - c*c + d*d;
        
    return rotation;
}

// renvoie la matrice (out) de transformation inverse de la matrice présente (*this)
// actuellement on recopie this
Transformation Transformation::inverse()
{
    Transformation out;
    out.rotation = this->rotation.transpose();
    out.position = -(out.rotation * this->position);
    return out;
}

// renvoie la matrice (out) de transformation comme la multiplication de la transformation A et de la transformation B
// en exemple on recopie A
Transformation operator* (const Transformation& A, const Transformation &B)
{
    Transformation out;
    out.rotation = A.rotation * B.rotation;
    out.position = A.rotation * B.position + A.position;
    return out;
}

// retourne la matrice de rotation en fonction des angles roll, pitch, yaw
Eigen::Matrix<double,3,3> setRPY( const double & roll, const double & pitch, const double & yaw)
{
    Eigen::Matrix<double,3,3> rotation;

    const double cr = cos(roll); // ROLL
    const double sr = sin(roll);
    const double cp = cos(pitch); // PITCH
    const double sp = sin(pitch);
    const double cy = cos(yaw); // YAW
    const double sy = sin(yaw);

    rotation(0,0) = cy * cp;
    rotation(0,1) = cy * sp * sr - sy * cr;
    rotation(0,2) = cy * sp * cr + sy * sr;

    rotation(1,0) = sy * cp;
    rotation(1,1) = sy * sp * sr + cy * cr;
    rotation(1,2) = sy * sp * cr - cy * sr;

    rotation(2,0) = -sp;
    rotation(2,1) = cp * sr;
    rotation(2,2) = cp * cr;
    
    return rotation;
}

// retourne les angles roll, pitch et yaw à partir d'une matrice de rotation
Eigen::Matrix<double,3,1> getRPY(const Eigen::Matrix<double,3,3> mat)
{
    Eigen::Matrix<double,3,1> out;   
    out(1) = atan2(-mat(2,0), sqrt(mat(0,0)*mat(0,0) + mat(1,0)*mat(1,0)));
    out(0) = atan2(mat(2,1), mat(2,2));
    out(2) = atan2(mat(1,0), mat(0,0));
    
    return out;
}


// Renvoie la matrice de transformation pour une rotation autour de X
Transformation RotX(double q)
{
    double s = sin(q);
    double c = cos(q);
    Transformation out;
    out.rotation(0,0) = 1;
    out.rotation(0,1) = 0;
    out.rotation(0,2) = 0;
    out.rotation(1,0) = 0;
    out.rotation(1,1) = c;
    out.rotation(1,2) = -s;
    out.rotation(2,0) = 0;
    out.rotation(2,1) = s;
    out.rotation(2,2) = c;
    
    return out;
}

// Renvoie la matrice de transformation pour une rotation autour de Y
Transformation RotY(double q)
{
    double s = sin(q);
    double c = cos(q);    
    Transformation out;
   out.rotation(0,0) = c;
    out.rotation(0,1) = 0;
    out.rotation(0,2) = s;
    out.rotation(1,0) = 0;
    out.rotation(1,1) = 1;
    out.rotation(1,2) = 0;
    out.rotation(2,0) = -s;
    out.rotation(2,1) = 0;
    out.rotation(2,2) = c;
    
    return out;   
}

// Renvoie la matrice de transformation pour une rotation autour de Z
Transformation RotZ(double q)
{
    double s = sin(q);
    double c = cos(q);    
    Transformation out;
    out.rotation(0,0) = c;
    out.rotation(0,1) = -s;
    out.rotation(0,2) = 0;
    out.rotation(1,0) = s;
    out.rotation(1,1) = c;
    out.rotation(1,2) = 0;
    out.rotation(2,0) = 0;
    out.rotation(2,1) = 0;
    out.rotation(2,2) = 1;

    return out;    
}

